aioshelly/wsrpc.py

```diff
@@ -0,0 +1,194 @@
+        if self.connected:
+            _LOGGER.debug("%s already connected", self._ip_address)
+            return
+
+        _LOGGER.debug("Trying to connect to device at %s", self._ip_address)
+        try:
+            self._websocket = await aiohttp_session.ws_connect(
+                f"http://{self._ip_address}/rpc", heartbeat=WS_HEARTBEAT_SEC
+            )
+        except (
+            client_exceptions.WSServerHandshakeError,
+            client_exceptions.ClientError,
+        ) as err:
+            raise CannotConnect(f"Error connecting to {self._ip_address}") from err
+
+        self._rx_task = asyncio.create_task(self._rx_msgs())
+
+        _LOGGER.info("Connected to %s", self._ip_address)
+
+    async def disconnect(self):
+        """Disconnect all sessions."""
+        if self._websocket is None:
+            return
+
+        websocket, self._websocket = self._websocket, None
+        await websocket.close()
+
+        for call_item in self._calls.items():
+            call_item.future.cancel()
+
+        self._calls = {}
+        self._rx_task = None
+
+    async def _handle_call(self, frame_id):
+        await self._websocket.send_json(
+            {
+                "id": frame_id,
+                "src": self._src,
+                "error": {"code": 500, "message": "Not Implemented"},
+            }
+        )
+
+    async def _handle_frame(self, frame):
+        if peer_src := frame.get("src", None):
+            if self._dst is not None and peer_src != self._dst:
+                _LOGGER.warning("Remote src changed: %s -> %s", self._dst, peer_src)
+            self._dst = peer_src
+
+        frame_id = frame.get("id", None)
+        method = frame.get("method", None)
+
+        if method:
+            # peer is invoking a method
+            params = frame.get("params", None)
+            if frame_id:
+                # and expects a response
+                _LOGGER.debug("handle call for frame_id: %s", frame_id)
+                await self._handle_call(frame_id)
+            else:
+                # this is a notification
+                _LOGGER.debug("Notification: %s %s", method, params)
+                await self._on_notification(method, params)
+
+        elif frame_id:
+            # looks like a response
+            if frame_id not in self._calls:
+                _LOGGER.warning("Response for an unknown request id: %s", frame_id)
+                return
+
+            call = self._calls.pop(frame_id)
+            call.resolve.set_result(frame)
+
+        else:
+            _LOGGER.warning("Invalid frame: %s", frame)
+
+    async def _rx_msgs(self):
+        async for msg in self._websocket:
+            _LOGGER.debug("Received Message: Type: %s,  Data: %s", msg.type, msg.data)
+
+            try:
+                frame = msg.json()
+            except ValueError as err:
+                raise InvalidMessage("Received invalid JSON.") from err
+            else:
+                await self._handle_frame(frame)
```

Comment:
This should not be async. By making it async you now have to wait for any callback to finish before you process the next message, instead of being able to process the 2nd message while the 1st message is processed in the background etc. Preferably you keep `_handle_frame` fully non-async, and let the listeners of the future decide if they want to be async.